### PR TITLE
Fuse aggregator

### LINKED ISSUE #1965: Add shape aggregator based on fuse
For data shape exploration, it is useful to be able to view the union of type descriptors for a set of records.  For example, when coming up with the json type schema for suricata logs, I wrote hacky one-off scripts to do the equivalent of 
 what `fuse (typeof(.)) by alert.signature, alert.category`. Maybe the aggregator will be called `shape` rather than `fuse`.

@mccanne prototype this on a branch, but it had bug https://github.com/brimsec/zq/issues/1959 and stayed unmerged. Once we fix #1959, the aggregator should follow easily.

### PR DESCRIPTION
**Pull request by henridf**
This PR adds the `fuse()` aggregator (with a placeholder name per https://github.com/brimsec/zq/issues/2114). 

Upon encountering collisions (columns with same name but different types), it computes the type union of the types rather than the `_n` suffixes that are used by the fuse proc. 

With this it should be straightforward to add "null absorption" (@mccanne 's "nullify"); we'll have to decide if that should be the default or something else.

close #1965

### CODE REVIEW

**Comment by henridf** [comment]
rebased to master with @mccanne's recent changes which allowed reverting the "parsing hack", and added a test for partials that I had forgotten to `git add`.

File: expr/agg/fuse.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	schema.unify = true
+
+	tt := zson.NewTypeTable(zctx)
+	for _, p := range f.partials {
+		typ, err := tt.LookupType(string(p.Bytes))
+		if err != nil {
+			return zng.Value{}, fmt.Errorf("invalid partial value: %s", err)
+		}
+		recType, isRecord := typ.(*zng.TypeRecord)
+		if !isRecord {
```
Comment (nwt) [inline expr/agg/fuse.go:48]:
```suggestion
		recType, ok := typ.(*zng.TypeRecord)
		if !ok {
```

File: expr/agg/schema.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+func findColByName(cols []zng.Column, name string) (zng.Column, int, bool) {
+	for i, col := range cols {
+		if col.Name == name {
+			return col, i, true
+		}
+	}
+	return zng.Column{}, -1, false
+}
+
```
Comment (nwt) [inline expr/agg/schema.go:55]: This is unused.
```suggestion
```

File: zql/zql.peg [hunk 3]
```diff
@@ -390,8 +390,14 @@ RenameProc
       RETURN(MAP("op": "RenameProc", "fields": PREPEND(first, rest)))
     }
 
+
+
```
Comment (nwt) [inline zql/zql.peg:394]: Nit: We usually have just one blank line between rules.
```suggestion
```

File: expr/agg/schema.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+			fused[i] = zng.Column{acol.Name, nest}
+		default:
+			if s.unify {
+				fused[i] = zng.Column{acol.Name, unify(s.zctx, acol.Type, bcol.Type)}
+			} else {
+				dis := disambiguate(fused, acol.Name)
+				renames.Srcs = append(renames.Srcs, append(path, acol.Name))
+				renames.Dsts = append(renames.Dsts, append(path, dis))
+				fused = append(fused, zng.Column{dis, bcol.Type})
+			}
```
Comment (nwt) [inline expr/agg/schema.go:137]:
```suggestion
		case s.unify:
			fused[i] = zng.Column{acol.Name, unify(s.zctx, acol.Type, bcol.Type)}
		default:
			dis := disambiguate(fused, acol.Name)
			renames.Srcs = append(renames.Srcs, append(path, acol.Name))
			renames.Dsts = append(renames.Dsts, append(path, dis))
			fused = append(fused, zng.Column{dis, bcol.Type})
```

**Review by nwt — APPROVED** [review]

File: expr/agg/fuse.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"fmt"
+
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zson"
+)
+
+type Fuse struct {
```
[the remark this answers is not in the document]
Comment (henridf) [reply expr/agg/fuse.go:12]: Probably not, changed. We do have a lot of capitalized types and structs that are currently package-local.  

It does make declarations look more important 🤪 .

File: expr/agg/fuse.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	f.shapes[typ] = struct{}{}
+	return nil
+}
+
+func (f *Fuse) Result(zctx *resolver.Context) (zng.Value, error) {
+	if len(f.shapes)+len(f.partials) == 0 {
+		// empty input
+		return zng.Value{zng.TypeNull, nil}, nil
+	}
+	schema, _ := NewSchema(zctx)
```
[the remark this answers is not in the document]
Comment (henridf) [reply expr/agg/fuse.go:38]: oops

File: expr/agg/schema.go (hunk 2 above)
Comment (henridf) [reply expr/agg/schema.go:55]: removed

**Pull request merged**